### PR TITLE
Update `eslint-plugin-jsdoc` to support Node 18 and 19

### LINKED ISSUE #45696: `@wordpress/eslint-plugin` package does not support node.js 18
### Description

Node.js 18 is LTS now.
But `@wordpress/eslint-plugin` dependency `eslint-plugin-jsdoc` version is too old.

Error:
```
error eslint-plugin-jsdoc@37.9.7: The engine "node" is incompatible with this module. Expected version "^12 || ^14 || ^16 || ^17". Got "18.12.1"
```

Command `yarn why eslint-plugin-jsdoc` output:
```
=> Found "eslint-plugin-jsdoc@37.9.7"
info Reasons this module exists
   - "@wordpress#scripts#@wordpress#eslint-plugin" depends on it
   - Hoisted from "@wordpress#scripts#@wordpress#eslint-plugin#eslint-plugin-jsdoc"
```

### Step-by-step reproduction instructions

1. Install Node.js 18
2. `yarn add @wordpress/scripts`

### Screenshots, screen recording, code snippet

_No response_

### Environment info

Node version: 18.12.1
@wordpress/scripts version: 24.5.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

No

### PR DESCRIPTION
**Pull request by herndlm**
Based on https://github.com/WordPress/gutenberg/pull/45697, credits go to @Antonio-Laguna

## What?

This updates `eslint-plugin-jsdoc` which is part of `eslint-plugin` to a version that allows Node 18 and 19.

## Why?

Right now installing these packages on Node 18 results in errors due to that plugin requiring a previous version of Node.

## How?

The package has been updated.

Supersedes https://github.com/WordPress/gutenberg/pull/45697
Fixes https://github.com/WordPress/gutenberg/issues/45696

I'll create the changelog in a minute, theoretically it would be a breaking change I guess, see
- https://github.com/gajus/eslint-plugin-jsdoc/releases/tag/v38.0.0 (match-description regular expressions now need to take account for trailing whitespace)
- https://github.com/gajus/eslint-plugin-jsdoc/releases/tag/v39.0.0 (drop Node 12 from engines, this one shouldn't matter here since Node 12 is not supported anymore)

### CODE REVIEW

**Comment by herndlm** [comment]
OK, I'm going to stop right here and now before I make this even worse. Please somebody let me know how I should fix the JSDoc types. E.g. a good example is `canMoveBlock()`. It had a return type of `boolean` but it has an internal sanity check that just exits with `return;` in case the block attributes are null. And that would be `undefined` instead of `boolean` which would be reported now (presumably because the `eslint-plugin-jsdoc` package got smarter)

There are multiple ways to adapt this I guess, but I can't decide this on my own, e.g.
- Adapt the JSDoc (+ docs) to `boolean|undefined`
- `return false;` if the block attributes are null
- raise an Error if the block attributes are null
- ..

also sorry for the big diff and review noise. this is the worst first PR ever :D feel free to let me know that this is a mess and should be done different. no harm done here, except a bit time lost. that's totally fine, I should have reached out before changing that many things :) UPDATE: tried to clean up the big diff and remove the unnecessary changes

**Comment by herndlm** [comment]
Added another commit that should fix the JSDoc issues. I also ran `lint:js:fix` to get rid of new alignment warnings that popped up, but realised too late how many files that changed and that I should have split this most likely, sorry :/

There were many incorrect return types, e.g. often `|undefined` was missing, but often also `?T` was used instead of `T|undefined`. So it could, most likely, also be fixed by doing `return null` instead of `return`. I'm not sure what effects these changes have, widening the return type is not nice for consumers I suppose (oh, the first e2e test already failed..) :/

Also: the commit hook complained with 
```
 Some API docs may be out of date: docs/reference-guides/data/data-core-block-editor.md,docs/reference-guides/data/data-core-editor.md,docs/reference-guides/data/data-core.md,packages/block-editor/README.md,packages/blocks/README.md,packages/core-data/README.md,packages/plugins/README.md,packages/shortcode/README.md Either build and stage them with npm run docs:build or continue with --no-verify.
```

which I still ignored. I want to get a first review here before further looking into this and fiddling around manually with types/docs..

**Comment by herndlm** [comment]
OK, I'm going to stop right here and now before I make this even worse. Please somebody let me know how I should fix the JSDoc types. E.g. a good example is `canMoveBlock()`. It had a return type of `boolean` but it has an internal sanity check that just exits with `return;` in case the block attributes are null. And that would be `undefined` instead of boolean.

There are multiple ways to adapt this I guess, but I can't decide this on my own, e.g.
- Adapt the JSDoc (+ docs) to `boolean|undefined`
- `return false;` if the block attributes are null
- raise an Error if the block attributes are null
- ..

File: packages/components/src/popover/utils.ts [hunk 1]
```diff
@@ -143,8 +143,8 @@ export const placementToMotionAnimationProps = (
  *
  * @param  document The iframe's owner document.
  *
- * @return The offset of the document's frame element, or undefined if the
- * document has no frame element.
+ * @return {Object|undefined} The offset of the document's frame element, or undefined if the
+ *                            document has no frame element.
```
Comment (herndlm) [inline packages/components/src/popover/utils.ts:147]: no idea why it started to complain here but not with the other functions in that file :/

**Comment by Antonio-Laguna** [comment]
@herndlm do you know if there's anything missing here?

**Comment by Antonio-Laguna** [comment]
@herndlm can you try to solve conflicts?

**Comment by herndlm** [comment]
sorry, without any feedback and not much of a clue how to handle type-related things here I'll close this for now

**Pull request closed without merging**